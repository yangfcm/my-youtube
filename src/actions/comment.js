import { defAxios as axios, maxResults } from "../settings";
import {
  ADD_COMMENT,
  FETCH_COMMENTS,
  FETCH_REPLIES,
  FETCH_COMMENTS_DISABLED,
  CATCH_ERROR,
} from "./types";
import { DEFAULT_ERROR_MSG } from "./default-error-msg";

/**
 * Fetch comments under a video
 * If accessToken is provided, the comments by authorized user will be displayed on the top.
 */
export const fetchComments = (
  videoId,
  pageToken,
  accessToken,
  order = "relevance"
) => {
  return async (dispatch) => {
    try {
      const response = await axios.get("/commentThreads", {
        headers: {
          Authorization: accessToken,
        },
        params: {
          part: "snippet",
          videoId,
          maxResults,
          pageToken,
          order: order === "relevance" ? "relevance" : "time",
        },
      });
      dispatch({
        type: FETCH_COMMENTS,
        payload: response.data,
      });
    } catch (e) {
      if (!e.response) {
        dispatch({
          type: CATCH_ERROR,
          payload: DEFAULT_ERROR_MSG.FAILED_TO_FETCH_COMMENTS,
        });
        return;
      }
      if (e.response.data.error.errors[0].reason === "commentsDisabled") {
        // Comment is disabled.
        dispatch({
          type: FETCH_COMMENTS_DISABLED,
          payload: DEFAULT_ERROR_MSG.COMMENT_DISABLED,
        });
      } else {
        dispatch({
          type: CATCH_ERROR,
          payload: e.response.data.error,
        });
      }
    }
  };
};

/** Add a comment to video */
export const addComment = (channelId, videoId, commentText, accessToken) => {
  return async (dispatch) => {
    try {
      const requestBdody = {
        snippet: {
          channelId,
          videoId,
          topLevelComment: {
            snippet: {
              textOriginal: commentText,
            },
          },
        },
      };
      const response = await axios.post("/commentThreads", requestBody, {
        headers: {
          Authorization: accessToken,
        },
        params: {
          part: "snippet",
        },
      });
      dispatch({
        type: ADD_COMMENT,
        payload: response.data,
      });
    } catch (e) {
      dispatch({
        type: CATCH_ERROR,
        payload: e.response.data.error,
      });
    }
  };
};

/** Fetch replies to a comment */
export const fetchCommentReplies = (commentId, pageToken) => {
  return async (dispatch) => {
    try {
      const response = await axios.get("/comments", {
        params: {
          part: "snippet",
          parentId: commentId,
          maxResults,
          pageToken,
        },
      });
      dispatch({
        type: FETCH_REPLIES,
        payload: response.data,
      });
    } catch (e) {
      dispatch({
        type: CATCH_ERROR,
        payload: e.response.data.error,
      });
    }
  };
};
