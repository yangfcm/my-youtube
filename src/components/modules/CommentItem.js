import React from "react";
import moment from "moment";

const CommentItem = ({ comment }) => {
  return (
    <React.Fragment>
      <div className="d-flex mb-2" style={{ overflow: "auto" }}>
        <div className="mr-3">
          <img
            src={comment.authorProfileImageUrl}
            alt={comment.authorDisplayName}
            className="rounded-circle"
          />
        </div>
        <div>
          <div className="mb-1">
            <span className="font-weight-bold mr-2">
              {comment.authorDisplayName}
            </span>
            <span className="text-muted">
              {moment(comment.publishedAt).fromNow()}
            </span>
          </div>
          <div dangerouslySetInnerHTML={{ __html: comment.textDisplay }}></div>
        </div>
      </div>
    </React.Fragment>
  );
};

export default CommentItem;
