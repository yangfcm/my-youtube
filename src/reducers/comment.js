import {
  ADD_COMMENT,
  FETCH_COMMENTS,
  FETCH_REPLIES,
  FETCH_COMMENTS_DISABLED,
} from "../actions/types";

export const commentReducer = (state = { isDisabled: false }, action) => {
  switch (action.type) {
    case FETCH_COMMENTS:
      return {
        ...state,
        comments: action.payload,
        isDisabled: false,
      };
    case FETCH_COMMENTS_DISABLED:
      return {
        ...state,
        isDisabled: true,
        comments: null,
      };
    case FETCH_REPLIES:
      return {
        ...state,
        replies: action.payload,
      };
    case ADD_COMMENT:
      return {
        ...state,
        addedComment: action.payload,
      };
    default:
      return state;
  }
};
