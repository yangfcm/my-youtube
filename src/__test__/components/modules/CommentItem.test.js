import React from "react";
import { shallow } from "enzyme";
import CommentItem from "components/modules/CommentItem";
import { commentItem } from "__test__/fixtures/comment";

describe("Test CommentItem component", () => {
  let wrapper;

  beforeEach(() => {
    wrapper = shallow(<CommentItem comment={commentItem} />);
  });

  afterEach(() => {
    wrapper.unmount();
  });

  it("should render component", () => {
    expect(wrapper).toMatchSnapshot();
  });

  it("should render author's avatar", () => {
    expect(wrapper.find("img").prop("src")).toBe(
      commentItem.authorProfileImageUrl
    );
  });

  it("should render author's username", () => {
    expect(wrapper.find("span.font-weight-bold").text()).toContain(
      commentItem.authorDisplayName
    );
  });

  it("should render comment text", () => {
    expect(
      wrapper.find("div#comment-text").prop("dangerouslySetInnerHTML").__html
    ).toContain(commentItem.textDisplay);
  });
});
