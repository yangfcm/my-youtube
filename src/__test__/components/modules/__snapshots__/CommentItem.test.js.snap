// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test CommentItem component should render component 1`] = `
<Fragment>
  <div
    className="d-flex mb-2"
    style={
      Object {
        "overflow": "auto",
      }
    }
  >
    <div
      className="mr-3"
    >
      <img
        alt="Lucia Lakpor"
        className="rounded-circle"
        src="https://yt3.ggpht.com/a/AATXAJxSLxDsYIuligoT4HSSVXBAEKk-MbmjBa4O606L=s48-c-k-c0xffffffff-no-rj-mo"
      />
    </div>
    <div>
      <div
        className="mb-1"
      >
        <span
          className="font-weight-bold mr-2"
        >
          Lucia Lakpor
        </span>
        <span
          className="text-muted"
        >
          7 months ago
        </span>
      </div>
      <div
        dangerouslySetInnerHTML={
          Object {
            "__html": "She only wanted to be president, she and her children stole everything and left the country in poverty and in the hands of another criminal, we basically have no country to go to, Liberia is now the poorest in the world. Yes I said it sheâ€™s a criminal.",
          }
        }
        id="comment-text"
      />
    </div>
  </div>
</Fragment>
`;
